### PR TITLE
Authentication

### PR DESCRIPTION
**Pull request by benm93**
This branch adds the user log in and registration features and also makes easy helper methods available to make it easy to add access control to controllers.

### CODE REVIEW

File: app/controllers/users_controller.rb [hunk 1]
```diff
@@ -1,7 +1,23 @@
 class UsersController < ApplicationController
   def new
+    @user = User.new_anonymous_user
```
Comment (ghost) [inline app/controllers/users_controller.rb]: What's the purpose of this `new_anonymous_user` method?

File: app/helpers/sessions_helper.rb [hunk 2]
```diff
@@ -0,0 +1,27 @@
+ module SessionsHelper
+  # Logs in the given user.
+  def log_in(user)
+    session[:user_id] = user.email_hash
```
Comment (ghost) [inline app/helpers/sessions_helper.rb]: Should this be:

`session[:user_id] = user.id`

?

**Review by ghost — COMMENTED** [review]
So I had two comments, but this looks good as a start. My only substantial question is I see we use the email_hash to "log in", but shouldn't that be the user id? 
Also - when I tried it locally and logged in it displayed the user hash on the front end, is that intentional?

Otherwise - this looks good to me, I'd be interested to hear Ed's thoughts for the test coverage.

Good next steps:
- password complexity validation
- email input validation
- remove email hash display
- Can we log out? -> after logging in change "log in" to "sign out"
- we shouldn't display the user_id in the url

Nice things to have:
- Account activiation via email?
- Forget password email

File: app/helpers/sessions_helper.rb (hunk 2 above)
Comment (benm93) [reply app/helpers/sessions_helper.rb]: Yeah I'll update that

**Review by EJRyan89 — COMMENTED** [review]
Would it make more sense to do the hashing in the models or the controllers? The User model has email and password fields in memory (only) so that we can handle them unhashed, but we can get rid of those if they are not necessary.

As for the tests, it is a good start for now. I would want to look into doing testing to make sure that sessions are restricted to a single user going forward (and that the user can't tamper with them).

Also, if I change the ID in the query string once I have authenticated, I can impersonate any user I want. If we are okay with that for the demo, we can have that bug in there, but that really should get fixed before merging to master.

File: app/controllers/users_controller.rb (hunk 1 above)
Comment (benm93) [reply app/controllers/users_controller.rb]: I meant to remove that, it'll be gone in the next commit.

**Pull request merged**